packages/snaps-rpc-methods/src/permitted/setState.ts

```diff
@@ -0,0 +1,252 @@
+    return end(error);
+  }
+
+  return end();
+}
+
+/**
+ * Validate the parameters of the `snap_setState` method.
+ *
+ * @param params - The parameters to validate.
+ * @returns The validated parameters.
+ */
+function getValidatedParams(params?: unknown) {
+  try {
+    return create(params, SetStateParametersStruct);
+  } catch (error) {
+    if (error instanceof StructError) {
+      throw rpcErrors.invalidParams({
+        message: `Invalid params: ${error.message}.`,
+      });
+    }
+
+    /* istanbul ignore next */
+    throw rpcErrors.internal();
+  }
+}
+
+/**
+ * Get the new state of the Snap.
+ *
+ * If the key is `undefined`, the value is expected to be an object. In this
+ * case, the value is returned as the new state.
+ *
+ * If the key is not `undefined`, the value is set in the state at the key. If
+ * the key does not exist, it is created (and any missing intermediate keys are
+ * created as well).
+ *
+ * @param key - The key to set.
+ * @param value - The value to set the key to.
+ * @param encrypted - Whether the state is encrypted.
+ * @param getSnapState - The `getSnapState` hook.
+ * @returns The new state of the Snap.
+ */
+async function getNewState(
+  key: string | undefined,
+  value: Json,
+  encrypted: boolean,
+  getSnapState: SetStateHooks['getSnapState'],
+) {
+  if (key === undefined) {
+    assert(isObject(value));
+    return value;
+  }
+
+  const state = await getSnapState(encrypted);
+  return set(state, key, value);
+}
+
+/**
+ * Set the value of a key in an object. The key may contain Lodash-style path
+ * syntax, e.g., `a.b.c` (with the exception of array syntax). If the key does
+ * not exist, it is created (and any missing intermediate keys are created as
+ * well).
+ *
+ * This is a simplified version of Lodash's `set` function, but Lodash doesn't
+ * seem to be maintained anymore, so we're using our own implementation.
+ *
+ * @param object - The object to get the key from.
+ * @param key - The key to set.
+ * @param value - The value to set the key to.
+ * @returns The new object with the key set to the value.
+ */
+export function set(
+  // eslint-disable-next-line @typescript-eslint/default-param-last
+  object: Record<string, Json> | null,
+  key: string | undefined,
+  value: Json,
+): JsonObject {
+  if (key === undefined) {
+    assert(isObject(value));
+    return value;
+  }
+
+  const keys = key.split('.');
+  const requiredObject = object ?? {};
+  let currentObject: Record<string, Json> = requiredObject;
+
+  for (let i = 0; i < keys.length; i++) {
+    const currentKey = keys[i];
+    if (FORBIDDEN_KEYS.includes(currentKey)) {
+      throw rpcErrors.invalidParams(
+        'Invalid params: Key contains forbidden characters.',
+      );
+    }
+
+    if (i === keys.length - 1) {
+      currentObject[currentKey] = value;
+      return requiredObject;
+    }
+
+    currentObject[currentKey] = isObject(currentObject[currentKey])
+      ? currentObject[currentKey]
+      : {};
```

Comment:
I'm still a little confused by this, if `currentObject[currentKey]` is an object we are just re-assigning it to the same value? Couldn't we do this instead?
```suggestion
    if (!isObject(currentObject[currentKey])) {
      currentObject[currentKey] = {};
    }
```

Additionally, this would replace anything that's not an object with this key, are we sure about that? We could fail instead and only define the object if the key is empty?